File: 129.js
// Leetcode 129 Sum Root to Leaf Numbers 
// References:https://leetcode.com/problems/sum-root-to-leaf-numbers/solutions/5025151/sum-root-to-leaf-numbers-beated-of-the-users/?envType=study-plan-v2&envId=top-interview-150

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var sumNumbers = function(root) {
    let ans = 0;

    const dfs = (node, path) => {
        if (!node) return;
        if (!node.left && !node.right){
            ans += path * 10 + node.val;
            return;
        }
        dfs(node.left, path * 10 + node.val);
        dfs(node.right, path * 10 + node.val);
    }
    dfs(root, 0);
    return ans;
};

File: 129.java
// Leetcode 129 Sum Root to Leaf Numbers 
// References:https://leetcode.com/problems/sum-root-to-leaf-numbers/solutions/41531/clean-java-dfs-solution-preorder-traversal/?envType=study-plan-v2&envId=top-interview-150
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int total;

    public int sumNumbers(TreeNode root) {
        total = 0;
        helper(root, 0);
        return total;
    }

    void helper(TreeNode root, int sum){
        if (root == null) return;

        sum = sum * 10 + root.val;

        if (root.left == null && root.right == null){
            total += sum;
            return;
        }
        helper(root.left, sum);
        helper(root.right, sum);
    }
}